2420 upload fridge tag data

andreievg opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #2420 

# 👩🏻‍💻 What does this PR do? 

Add upload fridgetag button to top left of sensor list. When fridgetag file is selected it will be uploaded and then integrated.

It's branched from this PR: ~~https://github.com/openmsupply/open-msupply/pull/2421 please see description there~~ https://github.com/openmsupply/open-msupply/pull/2383.

This PR also not directly related changes:
* anyhow dependency moved to workspace 
* created prepare file dir utility, since I saw common functionality when looking up static file for reference

# 🧪 How has/should this change been tested? 

I used [the blank test data file](https://drive.google.com/drive/u/1/search?q=simple%20sync), turned on vaccine module preference for the store. Synced, then went to sensor list, pressed the fridge tag button and used this file: https://github.com/openmsupply/temperature-sensor/blob/main/data/FridgeTag%202L/130500109088_202206081014.txt

## 💌 Any notes for the reviewer?

Turn off white space diff, especially for the .rs files (a few indentation changes).

Cargo.lock file is due to adding multipart form for actix-web, btw, that's half of the diff

There is a bit of carry over, to be done in subsequent issues, but I think what we have here should be sufficient form MVP, thanks @adrianwb for majority of the work done with Berlinger Fridge tags.

Carry over issue here (TODO)

## 📃 Documentation
 
Instruction on how to upload berlinger fridge tag info and what fridgetags we support should be added to cold chain docs

andreievg commented on client/packages/coldchain/src/Sensor/ListView/AppBarButtons.tsx:
```diff
@@ -0,0 +1,75 @@
+      if (!result.ok) {
+        throw new Error(await result.text());
+      }
+      // TODO prettify
```
Errors and success just appears in the error and success modal, added to this issue: https://github.com/openmsupply/open-msupply/issues/2423

andreievg commented on client/packages/coldchain/src/Sensor/ListView/AppBarButtons.tsx:
```diff
@@ -0,0 +1,75 @@
+  };
+
+  return (
+    <AppBarButtonsPortal>
```
Not sure if it's in the right place and labeled correctly, also added to this issue: https://github.com/openmsupply/open-msupply/issues/2423

andreievg replied:
Change name here, thanks: b54860e17e8f5ac27f8560c9d2b43325d47f4c3e

andreievg commented on server/server/src/upload_fridge_tag.rs:
```diff
@@ -0,0 +1,78 @@
+    cfg.service(upload);
+}
+
+#[derive(Debug, MultipartForm)]
```
For form data (file)

andreievg commented on server/server/src/upload_fridge_tag.rs:
```diff
@@ -0,0 +1,78 @@
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case")]
```
For store-id passed through url params

andreievg commented on client/packages/coldchain/src/Sensor/ListView/AppBarButtons.tsx:
```diff
@@ -0,0 +1,75 @@
+
+    try {
+      const result = await fetch(
+        `${Environment.UPLOAD_FRIDGE_TAG}?store-id=${storeId}`,
```
Please note that store-id is passed to the operation here

andreievg commented on server/server/src/upload_fridge_tag.rs:
```diff
@@ -0,0 +1,78 @@
+    settings: Data<Settings>,
+    service_provider: Data<ServiceProvider>,
+) -> HttpResponse {
+    // TODO Permissions
```
Added to: https://github.com/openmsupply/open-msupply/issues/2423

andreievg commented on server/server/src/upload_fridge_tag.rs:
```diff
@@ -0,0 +1,78 @@
+    service_provider: &ServiceProvider,
+) -> anyhow::Result<ReadSensor> {
+    let ctx = service_provider
+        .basic_context()
```
could add store_id here I guess, but then have to change all of the service methods to use it, added to https://github.com/openmsupply/open-msupply/issues/2423

andreievg commented on server/server/src/upload_fridge_tag.rs:
```diff
@@ -0,0 +1,78 @@
+        .basic_context()
+        .context("Cannot get connection")?;
+
+    let file = form.files.pop().context("Cannot find attached file")?;
```
.context here is from anyhow

andreievg commented on server/service/src/sensor/berlinger.rs:
```diff
@@ -1,3 +1,4 @@
+use anyhow::Context;
```
I did some refactoring in this file, best viewed with white space diff off:
* Return type of ReadSensor (number of logs and breaches)
* Add ReadSensorError (makes it easier to cast and return error in main method)
* Read sensor method processes just one sensor

andreievg commented on server/service/src/static_files.rs:
```diff
@@ -1,8 +1,8 @@
 use std::io::Error;
 use std::path::PathBuf;
-use std::str::FromStr;
```
As per comment in main PR description, moved some common functionality to utils

andreievg commented on server/service/src/static_files.rs:
```diff
@@ -26,20 +26,14 @@ pub struct StaticFileService {
 
     pub fn store_file(&self, file_name: &str, bytes: &[u8]) -> anyhow::Result<StaticFile> {
         let id = uuid();
-        std::fs::create_dir_all(&self.dir)?;
```
This is done in the prepare_file_dir method now (when this struct is created)

andreievg commented on server/server/src/upload_fridge_tag.rs:
```diff
@@ -0,0 +1,78 @@
+    // Move file
+    std::fs::rename(file.file.path(), &new_file_path)?;
+
+    ctx.connection
```
Please note all integrations happen in transaction now (to avoid partial and incomplete data being integrate)

andreievg commented on server/service/src/sensor/berlinger.rs:
```diff
@@ -263,92 +266,95 @@ fn sensor_add_if_new(
+    Ok(result)
 }
+
+// TODO tests
```
Adde to this one: https://github.com/openmsupply/open-msupply/issues/2423

adrianwb reviewed (approved):
I don't know enough to properly review the client-side code, but the rest looks fine, apart from a typo ("worspace" instead of "workspace") -> I just pushed a commit to fix it rather than delay the review. I've also renamed `read_sensors` as `read_sensor` as it's now only reading a single sensor. Assuming you're ok with those, I'll leave you to merge it...

andreievg commented:
> I don't know enough to properly review the client-side code, but the rest looks fine, apart from a typo ("worspace" instead of "workspace") -> I just pushed a commit to fix it rather than delay the review. I've also renamed read_sensors as read_sensor as it's now only reading a single sensor. Assuming you're ok with those, I'll leave you to merge it...

Thanks @adrianwb.

I have a script mentioned in mSupply dev chat at some point, where I can quickly create a branch from current branch and open it for compare, so instead of committing straight to PR, it creates a suggestion link for an easy mergable branch: 

https://gist.github.com/andreievg/a091b81f5cc0ac8abd061588ab206bb8

It works quite well since most of the PR reviewing i do in VS code, where making change suggestions is quite easy

mark-prins reviewed (changes requested):
🎉  It's fantastic! 
A couple of small changes required:

- The `package/..` imports need to be fixed up
- The `migrations/v1_05_00/cold_chain.rs` needs to have

```
                ALTER TYPE sensor_type ADD VALUE IF NOT EXISTS 'BERLINGER';
```

added for postgres.

Apart from that, it's all good though I would also like 
- Button label would be better if it said what is going to happen (upload / import) rather than a noun (Fridge tags)
- the icon on the button
- the loading state looks weird on my screen with the white background on pale grey:
<img width="163" alt="Screenshot 2023-11-03 at 2 35 09 PM" src="https://github.com/openmsupply/open-msupply/assets/9192912/36f4a1eb-a91b-4b7b-927a-bc3ad3bdcfa1">

andreievg commented on client/packages/coldchain/src/Sensor/ListView/AppBarButtons.tsx:
```diff
@@ -0,0 +1,75 @@
+        style={{ display: 'none' }} // Make the file input element invisible
+      />
+      <LoadingButton
+        startIcon={<RadioIcon />}
```
Added to carry over, it has some prettification points, which I think this is a part of ?
https://github.com/openmsupply/open-msupply/issues/2423

andreievg commented:
@mark-prins thanks for suggestions, I committed your changes, renamed the button and added icon change to cary over, I think the migrations are fine ?

> ALTER TYPE sensor_type ADD VALUE IF NOT EXISTS 'BERLINGER';

I thought below is sufficient ?

https://github.com/openmsupply/open-msupply/blob/b54860e17e8f5ac27f8560c9d2b43325d47f4c3e/server/repository/src/migrations/v1_05_00/sensor.rs#L10-L20

mark-prins commented:
> I thought below is sufficient ?

🤦  yes, of course. I'm running v1.5 already soz

mark-prins reviewed (approved):
Thanks!
